### PR TITLE
Add new upgrades guide

### PR DESCRIPTION
**Pull request by spalladino**
Adds a new guide on upgrades, that covers CLI, writing upgradeable contracts, upgrades.js, and how upgrades work. I'm not sure whether to include this last section though.

Note that there may be a several links missing, we can make another pass and add them once we have settled on the final structure.

### CODE REVIEW

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 1]
```diff
@@ -1,8 +1,255 @@
...
+    // Reads the last stored value
+    function retrieve() public view returns (uint256) {
+        return value;
+    }
+}
+```
+
+NOTE: If you had stopped and restarted ganache between the xref:deploy-and-interact.adoc[deployment guide] and this one, you will need to first `create` a new `Box` instance, so you can now upgrade it.
+
+For the sopenzeppelin-sdkake of the example, let's say we want to add a new feature: a function that increments the `value` stored in the `Box`.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
For the sake of the example, let's say we want to add a new feature: a function that increments the `value` stored in the `Box`.
```

:joy:

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 2]
```diff
@@ -1,8 +1,255 @@
 = Upgrades
 
-Are smart contracts immutable?
-Hint: https://en.wikipedia.org/wiki/Betteridge%27s_law_of_headlines
+Smart contracts xref:interact.adoc[deployed] with the OpenZeppelin CLI can be **upgraded** to modify their code, while preserving their address, state, and balance. This allows you to iteratively add new features to your project, or fix any bugs you may find in xref:mainnet.adoc[in production].
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
Smart contracts xref:deploy-and-interact.adoc[deployed] with the OpenZeppelin CLI can be **upgraded** to modify their code, while preserving their address, state, and balance. This allows you to iteratively add new features to your project, or fix any bugs you may find in xref:mainnet.adoc[in production]. 
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 3]
```diff
@@ -1,8 +1,255 @@
...
+
+* <<upgrading-a-contract-via-cli, Upgrade our Box contract using the CLI>>
+* <<limitations-of-contracts-upgrades, Learn how to write upgradeable contracts>>
+* <<upgrading-contracts-in-js, Use upgrades.js to programmatically manage upgrades>>
+* <<how-upgrades-work, Learn how upgrades work behind the hood>>
+
+[[upgrading-a-contract-via-cli]]
+== Upgrading a contract using the CLI
+
+Whenever you deploy a new contract using the CLI via `openzeppelin create`, that contract instance can be **upgraded** later. By default, only the address that originally deployed the contract can upgrade it. Let's see how it works, by upgrading the `Box` contract xref:deploy-and-interact.adoc#deploying-a-smart-contract[we deployed earlier]:
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
Whenever you deploy a new contract using the CLI via `openzeppelin create`, that contract instance can be **upgraded** later. By default, only the address that originally deployed the contract can upgrade it. 

Let's see how it works, by upgrading the `Box` contract xref:deploy-and-interact.adoc#deploying-a-smart-contract[we deployed earlier]:
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 4]
```diff
@@ -1,8 +1,255 @@
...
+
+    // Increments the stored value by 1
+    function increment() public {
+        value = value + 1;
+        emit ValueChanged(value);
+    }
+}
+```
+
+After changing the Solidity file, we can now just upgrade the instance we had deployed earlier by running the `openzeppelin upgrade` command.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc:79]: I think this is a bit of an extreme case of "show, don't tell". Because of how easy it is to perform the upgrade, I worry the reader may not realize the significance of the actions performed. It reminds me of a scene in The Prestige :stuck_out_tongue: 

What do you think about adding a small snippet going over what changing the code would normally require (migrate state and funds, change all references to the old address)?

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 5]
```diff
@@ -1,8 +1,255 @@
...
+```bash
+$ npx openzeppelin upgrade
+? Pick a network: development
+✓ Compiled contracts with solc 0.5.9
+✓ Contract Box deployed
+? Which proxies would you like to upgrade?: All proxies
+Instance upgraded at 0xCfEB869F69431e42cdB54A4F4f105C19C080A601.
+```
+
+Done! Our `Box` instance has been upgraded to the latest version of the code, while keeping its state and the same address as before. Let's try it out by invoking the new `increment` function, and checking the `value` afterwards:
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
Done! Our `Box` instance has been upgraded to the latest version of the code, while keeping its state and the same address as before. 

Let's try it out by invoking the new `increment` function, and checking the `value` afterwards:
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 6]
```diff
@@ -1,8 +1,255 @@
...
-Short explanation on how it works plus links
-Small example using upgrades.js
\ No newline at end of file
+Throughout this guide, we will:
+
+* <<upgrading-a-contract-via-cli, Upgrade our Box contract using the CLI>>
+* <<limitations-of-contracts-upgrades, Learn how to write upgradeable contracts>>
+* <<upgrading-contracts-in-js, Use upgrades.js to programmatically manage upgrades>>
+* <<how-upgrades-work, Learn how upgrades work behind the hood>>
+
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc:28]:
```suggestion
TIP: If you're unfamiliar with the OpenZeppelin CLI, head to xref:deploy-and-interact.adoc#getting-started-with-the-cli[our introductory guide] first!

```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 7]
```diff
@@ -1,8 +1,255 @@
...
+? Select which function: retrieve()
+✓ Method 'retrieve()' returned: 6
+```
+
+That's it! This process is the same regardless of whether you are working on a local blockchain, a testnet, or the main network. 
+
+[[limitations-of-contracts-upgrades]]
+== Limitations of contracts upgrades
+
+Smart contracts are immutable by default, so there are some limitations when making them upgradeable. These come in two flavors: when first creating your contract, and when changing it.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]: The conclusion here feels a bit shoehorned, specially considering we haven't yet covered the details behind how upgrades work.

I'd simplify this paragraph by stating something to the effect of:

'While any smart contract can be made upgradeable, some restrictions of the Solidity language need to be worked around. These come up when writing both the initial contract and version we'll upgrade it to.'

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 8]
```diff
@@ -1,8 +1,255 @@
...
+
+=== Initialization
+
+When writing an upgradeable contract, you cannot declare a `constructor`. If you want to run any initialization code, the SDK provides a special `Initializable` base contract that allows you to tag a method as `initializer`, ensuring it can be run only once.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]: The need to introduce `upgrades` here to access `Initializable` seems to point in the direction of merging the proxies into Contracts.

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 9]
```diff
@@ -1,8 +1,255 @@
...
+? _admin (address): 0x90f8bf6a479f320ead074411a4b0e7944ea8c9c1
+✓ Setting everything up to create contract instances
+✓ Instance created at 0x2612Af3A521c2df9EAF28422Ca335b04AdF3ac66
+```
+
+For all practical purposes, the initializer acts as a constructor. However, keep in mind that, since it's a regular function, you need to manually call the initializers of all base contracts, if any.
+
+NOTE: In future versions of the OpenZeppelin SDK, the CLI will take care of automatically converting constructors into initializers, so you do not need to worry about this.
+
+To learn more about this and other caveats when writing upgradeable contracts, check out our Writing Upgradeable Contracts guide.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
To learn more about this and other caveats when writing upgradeable contracts, check out our xref:sdk::writing-contracts.adoc[Writing Upgradeable Contracts] guide.
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 10]
```diff
@@ -1,8 +1,255 @@
...
+    
+    // ...
+}
+```
+
+Fortunately, this limitation only affects state variables. You can change the contract's functions and events as you wish.
+
+NOTE: If you accidentally mess up with your contract's storage layout, the CLI will warn you when you try to upgrade.
+
+To learn more about this limitation, head over to the Modifying Your Contracts guide.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
To learn more about this limitation, head over to the xref:sdk::writing-contracts.adoc#modifying-your-contracts[Modifying Your Contracts] guide.
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 11]
```diff
@@ -1,8 +1,255 @@
...
+An upgrade then involves the following steps:
+
+. Deploy the new implementation contract.
+. Send a transaction to the proxy, that updates its implementation address to the new one.
+
+NOTE: You can have multiple proxies using the same implementation contract, so you can save gas using this pattern if you plan to deploy multiple copies of the same contract.
+
+Any user of the smart contract always interacts with the proxy, which never changes its address. This allows you to roll out an upgrade or fix a bug without requesting your users to change anything on their end - they just keep interacting with the same address as always.
+
+NOTE: If you want to learn more about how OpenZeppelin proxies work, check out our Upgrades Pattern guide.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
NOTE: If you want to learn more about how OpenZeppelin proxies work, check out our xref:sdk::pattern.adoc[Upgrades Pattern guide].
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc (hunk 4 above)
Comment (nventuro) [reply components/learn/modules/ROOT/pages/on-upgrades.adoc:79]: After rereading a couple times, I think said snippet should be an initial motivation section, explaining why upgrades are: 
 a) useful,
 b) important, and
 c) **hard** (unless you use the SDK!)

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 12]
```diff
@@ -1,8 +1,255 @@
...
+NOTE: If you accidentally mess up with your contract's storage layout, the CLI will warn you when you try to upgrade.
+
+To learn more about this limitation, head over to the Modifying Your Contracts guide.
+
+[[upgrading-contracts-in-js]]
+== Upgrading contracts programmatically
+
+If you want to create and upgrade contracts from your javascript code instead of via the command line, you can use the `@openzeppelin/upgrades.js` library instead of the CLI.
+
+NOTE: The CLI does not just manage contract upgrades, but also compilation, interaction, and source code verification. The `upgrade.js` library only takes care of creating and upgrading. The library also does not keep track of the contracts you have already deployed, nor runs any initializer or storage layout validations, as the CLI does.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]: >The library also does not keep track of the contracts you have already deployed, nor runs any initializer or storage layout validations

Are these expected to be eventually supported? They seem like major reasons to not use the lib.

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 13]
```diff
@@ -1,8 +1,255 @@
...
+```js
+const address = instance.options.address;
+const MyContractV1 = Upgrades.Contracts.getFromLocal('MyContractV1');
+await project.upgradeProxy(address, MyContractV1);
+```
+
+That's it! You can now programmatically manage your contracts upgrades from your javascript code.
+
+[[how-upgrades-work]]
+== How upgrades work
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]: I was a bit surprised by this section, I would've expected for it to be its own article wherever the proxy contracts live (currently upgrades.js, possibly Contracts in the future). I'm not against having it here however, but I'd consider placing it before the 'limitations' section. 

Currently, writing upgradeable contracts, using the CLI and upgrades.js all require some level of understanding of the proxy pattern: the name even shows up when using these tools. Once the transpiler is ready and the programmatic interface redesigned this will become a technical detail, but I'm not sure we can sweep it under the rug for now.

**Review by nventuro — COMMENTED** [review]
Great job! This really fits with the theme of the learn guides, I'm surprised we didn't have this sort of content before.

I left a mix of nitpicky and overall structure comments: I didn't want to go into too much detail until we've settled on the general layout of the guide.

File: components/learn/modules/ROOT/pages/on-upgrades.adoc (hunk 1 above)
Comment (spalladino) [reply components/learn/modules/ROOT/pages/on-upgrades.adoc]: Dafuq?

File: components/learn/modules/ROOT/pages/on-upgrades.adoc (hunk 8 above)
Comment (spalladino) [reply components/learn/modules/ROOT/pages/on-upgrades.adoc]: Indeed. That's something I wanted to discuss with you and @frangio, but I think that OZ Contracts v3 should include Initializable and Proxies (and potentially even the App interface), as well as the migration to Solidity 0.6.

**Comment by spalladino** [comment]
@nventuro I've just changed the guide according to your suggestions. I added a short intro explaining the value of upgrades (and the difficulty of doing them by hand) and moved the "how they work" section to before the limitations one. I also added a few other changes here and there. Please re-review!

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 14]
```diff
@@ -1,8 +1,280 @@
...
\ No newline at end of file
+Throughout this guide, we will:
+
+* <<upgrading-a-contract-via-cli, Upgrade our Box contract using the CLI>>
+* <<limitations-of-contracts-upgrades, Learn how to write upgradeable contracts>>
+* <<upgrading-contracts-in-js, Use upgrades.js to programmatically manage upgrades>>
+* <<how-upgrades-work, Learn how upgrades work behind the hood>>
+
+[[what-is-an-upgrade]]
+== What's in an upgrade
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc:14]: I'm not sure if this is a typo, given that the identifier for the section reads 'what is an upgrade'. If not, I'd consider using simpler language (of the sort of 'why use upgradeable contracts'): the current phrasing might be hard to understand for non-native speakers.

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 15]
```diff
@@ -1,8 +1,280 @@
...
+
+Smart contracts in Ethereum are immutable by default. This means that, once you have created them, there is no way to alter them, effectively acting as an inalterable contract among participants.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
Smart contracts in Ethereum are immutable by default. Once you create them there is no way to alter them, effectively acting as an unbreakable contract among participants.
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 16]
```diff
@@ -1,8 +1,280 @@
...
+
+However, for some scenarios, it is desirable to be able to modify them. Think of a contract between two parties: if they both agree to change it, they should be able to do so. This could happen because they detected a bug they need to fix (which could even lead to a hacker stealing you funds!), they want to add additional features, or simply because the rules of the game changed.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
However, for some scenarios, it is desirable to be able to modify them. Think of a traditional contract between two parties: if they both agreed to change it, they would be able to do so. On Ethereum, they may desire to alter a smart contract to fix a bug they found (which might even lead to a hacker stealing their funds!), to add additional features, or simply to change the rules enforced by it.
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 17]
```diff
@@ -1,8 +1,280 @@
...
+
+Without being able to upgrade your contracts, if you spot a bug in one of them, you will need to:
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
Here's what you'd need to do to fix a bug in a contract you cannot upgrade:
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 18]
```diff
@@ -1,8 +1,280 @@
...
+
+. Deploy a new version of the contract
+. Manually migrate all state from the old one contract to the new one (which can be very expensive in terms of gas fees!)
+. Update all contracts that interacted with the old contract to use the address of the new one
+. Reach out to all your users and convince them to start using the new deployment (and handle both contracts being used simultaneously, as users are slow to migrate)
+
+To avoid going through this mess, we have built contract upgrades directly in our SDK. This allows us to *change the contract code, while preserving the state, balance, and address*. An upgrade then only requires running a single command. Let's see it in action.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
To avoid going through this mess, we have built contract upgrades directly in our SDK. This allows us to *change the contract code, while preserving the state, balance, and address*. And all by running a single command. Let's see it in action.
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 19]
```diff
@@ -1,8 +1,280 @@
...
+```bash
+$ npx openzeppelin upgrade
+? Pick a network: development
+✓ Compiled contracts with solc 0.5.9
+✓ Contract Box deployed
+? Which proxies would you like to upgrade?: All proxies
+Instance upgraded at 0xCfEB869F69431e42cdB54A4F4f105C19C080A601.
+```
+
+Done! Our `Box` instance has been upgraded to the latest version of the code, while keeping its state and the same address as before. We haven't had to manually deploy a new one at a new address, nor manually copy the `value` from the old `Box` to the new one.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
Done! Our `Box` instance has been upgraded to the latest version of the code, *while keeping its state and the same address as before*. We didn't need to deploy a new one at a new address, nor manually copy the `value` from the old `Box` to the new one.
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 20]
```diff
@@ -1,8 +1,280 @@
...
+```
+
+After changing the Solidity file, we can now just upgrade the instance we had deployed earlier by running the `openzeppelin upgrade` command.
+
+```bash
+$ npx openzeppelin upgrade
+? Pick a network: development
+✓ Compiled contracts with solc 0.5.9
+✓ Contract Box deployed
+? Which proxies would you like to upgrade?: All proxies
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
? Which instances would you like to upgrade?: All instances
```

This is the output on v2.6.0.

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 21]
```diff
@@ -1,8 +1,280 @@
...
+? Select which function: retrieve()
+✓ Method 'retrieve()' returned: 6
+```
+
+That's it! Notice how the `value` of the `Box` was preserved throughout the upgrade, as well as its address. And this process is the same regardless of whether you are working on a local blockchain, a testnet, or the main network. Let's see how the OpenZeppelin SDK accomplishes this.
+
+[[how-upgrades-work]]
+== How upgrades work
+
+_This section will be more theory-heavy than others, so feel free to skip over it, and return later if you are curious._
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
_This section will be more theory-heavy than others: feel free to skip over it and return later if you are curious._
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 22]
```diff
@@ -1,8 +1,280 @@
...
+
+When you create a new upgradeble contract instance, the OpenZeppelin SDK actually deploys two contracts:
+
+. The contract you have written, which is known as the _implementation contract_ or _logic contract_.
+. A _proxy_ to that contract, which is the instance that you actually interact with.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
. A _proxy_ to that contract, which is the contract that you actually interact with.
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 23]
```diff
@@ -1,8 +1,280 @@
...
+
+Here, the _proxy_ is a simple contract that just _delegates_ all calls to an implementation contract. A _delegate call_ is similar to a regular call, but all code is executed in the context of the caller, not of the callee. This means that a `transfer` in the implementation contract's code will actually transfer the proxy's balance. Similarly, any reads or writes to the contract storage will read or write from the proxy's storage.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
Here, the _proxy_ is a simple contract that just _delegates_ all calls to an implementation contract. A _delegate call_ is similar to a regular call, except that all code is executed in the context of the caller, not of the callee. Because of this, a `transfer` in the implementation contract's code will actually transfer the proxy's balance, and any reads or writes to the contract storage will read or write from the proxy's own storage.
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 24]
```diff
@@ -1,8 +1,280 @@
...
+
+This allows us to **decouple** a contract's state and code: the proxy holds the state, while the logic contract provides the code. And it also allows us to **change** the code by just having the proxy delegate to a different implementation contract.
+
+An upgrade then involves the following steps:
+
+. Deploy the new implementation contract.
+. Send a transaction to the proxy, that updates its implementation address to the new one.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
. Send a transaction to the proxy that updates its implementation address to the new one.
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 25]
```diff
@@ -1,8 +1,280 @@
...
+
+NOTE: You can have multiple proxies using the same implementation contract, so you can save gas using this pattern if you plan to deploy multiple copies of the same contract.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc:131]:
```suggestion
```

This is unrelated to upgrades, and I worry it will confuse people at the critical point of understanding.

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 26]
```diff
@@ -1,8 +1,280 @@
...
+
+Any user of the smart contract always interacts with the proxy, which never changes its address. This allows you to roll out an upgrade or fix a bug without requesting your users to change anything on their end - they just keep interacting with the same address as always.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
Any user of the smart contract always interacts with the proxy, *which never changes its address*. This allows you to roll out an upgrade or fix a bug without requesting your users to change anything on their end - they just keep interacting with the same address as always.
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 27]
```diff
@@ -1,8 +1,280 @@
...
+
+NOTE: If you want to learn more about how OpenZeppelin proxies work, check out our xref:sdk::pattern.adoc[Upgrades Pattern guide].
+
+[[limitations-of-contracts-upgrades]]
+== Limitations of contracts upgrades
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
== Limitations of Contract Upgrades
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 28]
```diff
@@ -1,8 +1,280 @@
...
+. Deploy the new implementation contract.
+. Send a transaction to the proxy, that updates its implementation address to the new one.
+
+NOTE: You can have multiple proxies using the same implementation contract, so you can save gas using this pattern if you plan to deploy multiple copies of the same contract.
+
+Any user of the smart contract always interacts with the proxy, which never changes its address. This allows you to roll out an upgrade or fix a bug without requesting your users to change anything on their end - they just keep interacting with the same address as always.
+
+NOTE: If you want to learn more about how OpenZeppelin proxies work, check out our xref:sdk::pattern.adoc[Upgrades Pattern guide].
+
+[[limitations-of-contracts-upgrades]]
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
[[limitations-of-contract-upgrades]]
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 29]
```diff
@@ -1,8 +1,281 @@
...
+NOTE: You can have multiple proxies using the same implementation contract, so you can save gas using this pattern if you plan to deploy multiple copies of the same contract.
+
+Any user of the smart contract always interacts with the proxy, which never changes its address. This allows you to roll out an upgrade or fix a bug without requesting your users to change anything on their end - they just keep interacting with the same address as always.
+
+NOTE: If you want to learn more about how OpenZeppelin proxies work, check out our xref:sdk::pattern.adoc[Upgrades Pattern guide].
+
+[[limitations-of-contract-upgrades]]
+== Limitations of Contract Upgrades
+
+While any smart contract can be made upgradeable, some restrictions of the Solidity language need to be worked around. These come up when writing both the initial contract and version we'll upgrade it to.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
While any smart contract can be made upgradeable, some restrictions of the Solidity language need to be worked around. These come up when writing both the initial contract and the version we'll upgrade it to.
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 30]
```diff
@@ -1,8 +1,281 @@
...
+
+=== Initialization
+
+When writing an upgradeable contract, you cannot declare a `constructor`. If you want to run any initialization code, the SDK provides a special `Initializable` base contract that allows you to tag a method as `initializer`, ensuring it can be run only once.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
Upgradeable contracts cannot have a `constructor`. To help you run initialization code, the SDK provides the `Initializable` base contract that allows you to tag a method as `initializer`, ensuring it can be run only once.
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 31]
```diff
@@ -1,8 +1,281 @@
...
+? Pick a network: development
+✓ Contract AdminBox deployed
+? Call a function to initialize the instance after creating it? Yes
+? Select which function: initialize(_admin: address)
+? _admin (address): 0x90f8bf6a479f320ead074411a4b0e7944ea8c9c1
+✓ Setting everything up to create contract instances
+✓ Instance created at 0x2612Af3A521c2df9EAF28422Ca335b04AdF3ac66
+```
+
+For all practical purposes, the initializer acts as a constructor. However, keep in mind that, since it's a regular function, you need to manually call the initializers of all base contracts, if any.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
For all practical purposes, the initializer acts as a constructor. However, keep in mind that since it's a regular function, you will need to manually call the initializers of all base contracts (if any).
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 32]
```diff
@@ -1,8 +1,281 @@
...
+
+NOTE: In future versions of the OpenZeppelin SDK, the CLI will take care of automatically converting constructors into initializers, so you do not need to worry about this.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
NOTE: In future versions of the OpenZeppelin SDK, the CLI will take care of automatically converting constructors into initializers, so you won't need to worry about this.
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 33]
```diff
@@ -1,8 +1,281 @@
...
+
+To learn more about this and other caveats when writing upgradeable contracts, check out our xref:sdk::writing-contracts.adoc[Writing Upgradeable Contracts] guide.
+
+=== Upgrading
+
+Due to technical limitations, when you upgrade a contract to a new version, you cannot change the **storage layout** of that contract.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
Due to technical limitations, when you upgrade a contract to a new version you cannot change the **storage layout** of that contract.
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 34]
```diff
@@ -1,8 +1,281 @@
...
+. Send a transaction to the proxy, that updates its implementation address to the new one.
+
+NOTE: You can have multiple proxies using the same implementation contract, so you can save gas using this pattern if you plan to deploy multiple copies of the same contract.
+
+Any user of the smart contract always interacts with the proxy, which never changes its address. This allows you to roll out an upgrade or fix a bug without requesting your users to change anything on their end - they just keep interacting with the same address as always.
+
+NOTE: If you want to learn more about how OpenZeppelin proxies work, check out our xref:sdk::pattern.adoc[Upgrades Pattern guide].
+
+[[limitations-of-contract-upgrades]]
+== Limitations of Contract Upgrades
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc:194]: I'd put this after 'upgrading contracts programatically', given that one is very related to what we have just read (and we don't want to deliver bad news too soon :sweat_smile:)

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 35]
```diff
@@ -1,8 +1,281 @@
...
+
+NOTE: The CLI does not just manage contract upgrades, but also compilation, interaction, and source code verification. The `upgrade.js` library only takes care of creating and upgrading. The library also does not keep track of the contracts you have already deployed, nor runs any initializer or storage layout validations, as the CLI does. Nevertheless, these capabilities may be added to `upgrade.js` in the near future.
+
+Your first step will be to install the library in your project, and you will also probably want to install `web3`:
+
+```bash
+$ npm install @openzeppelin/upgrades.js web3
+```
+
+As in our previous guide on programmatically interacting with contracts, we will start with some boilerplate code to initialize a provider, as well as the upgrades library.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
As in our previous guide on xref:deploy-and-interact.adoc#interacting-programmatically[programmatically interacting with contracts], we will start with some boilerplate code to initialize a provider, as well as the Upgrades library.
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 36]
```diff
@@ -1,8 +1,281 @@
...
+Fortunately, this limitation only affects state variables. You can change the contract's functions and events as you wish.
+
+NOTE: If you accidentally mess up with your contract's storage layout, the CLI will warn you when you try to upgrade.
+
+To learn more about this limitation, head over to the xref:sdk::writing-contracts.adoc#modifying-your-contracts[Modifying Your Contracts] guide.
+
+[[upgrading-contracts-in-js]]
+== Upgrading Contracts Programmatically
+
+If you want to create and upgrade contracts from your JavaScript code instead of via the command line, you can use the `@openzeppelin/upgrades.js` library instead of the CLI.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
If you want to create and upgrade contracts from your JavaScript code instead of via the command line, you can use the *OpenZeppelin Upgrades* library instead of the CLI.
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 37]
```diff
@@ -1,8 +1,281 @@
...
+
+NOTE: The CLI does not just manage contract upgrades, but also compilation, interaction, and source code verification. The `upgrade.js` library only takes care of creating and upgrading. The library also does not keep track of the contracts you have already deployed, nor runs any initializer or storage layout validations, as the CLI does. Nevertheless, these capabilities may be added to `upgrade.js` in the near future.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
NOTE: The CLI does not just manage contract upgrades, but also compilation, interaction, and source code verification. The Upgrades library only takes care of creating and upgrading. The library also does not keep track of the contracts you have already deployed, nor runs any initializer or storage layout validations, as the CLI does. Nevertheless, these capabilities may be added to the Upgrades library in the near future.
```

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 38]
```diff
@@ -1,8 +1,281 @@
...
+async function main() {
+  // Set up web3 object, connected to the local development network, initialize the Upgrades library
+  const web3 = new Web3('http://localhost:8545');
+  Upgrades.ZWeb3.initialize(web3.currentProvider)
+}
+
+main();
+```
+
+NOTE: You can check out a full version of the code in this section in the https://github.com/OpenZeppelin/openzeppelin-sdk/tree/master/examples/upgrades-library[upgrades-library example of the SDK repository].
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]: This link leads nowhere. Did you mean https://github.com/OpenZeppelin/openzeppelin-sdk/tree/master/examples/lib-simple?

File: components/learn/modules/ROOT/pages/on-upgrades.adoc [hunk 39]
```diff
@@ -1,8 +1,281 @@
...
+
+All our code from now on will be part of the `main` function. Let's begin by creating a new `project`, to manage our upgradeable contracts.
+
+```js
+const [from] = await ZWeb3.accounts();
+const project = new ProxyAdminProject('MyProject', null, null, { from, gas: 1e6, gasPrice: 1e9 });
+```
+
+NOTE: The Upgrades library ships with three different flavours of projects: `SimpleProject`, `ProxyAdminProject`, and `AppProject`. We recommend using the `ProxyAdmin` one to begin with. You can learn more in the upgrades.js documentation.
```
Comment (nventuro) [inline components/learn/modules/ROOT/pages/on-upgrades.adoc]:
```suggestion
NOTE: The Upgrades library ships with three different flavours of projects: `SimpleProject`, `ProxyAdminProject`, and `AppProject`. We recommend using the `ProxyAdmin` one to begin with. You can learn more in the Upgrades documentation.
```

Comment (nventuro) [reply components/learn/modules/ROOT/pages/on-upgrades.adoc]: Link pending completion of https://github.com/OpenZeppelin/openzeppelin-sdk/issues/1286

**Review by nventuro — COMMENTED** [review]
Looks great! I love the new intro section and how the whole thing now reads.

The programmatic guide is the weakest bit, but I don't think we'll be able to do much about that until the API is revised.

File: components/learn/modules/ROOT/pages/on-upgrades.adoc (hunk 25 above)
Comment (spalladino) [reply components/learn/modules/ROOT/pages/on-upgrades.adoc:131]: It is related to proxies though. I agree it could disrupt the flow of the explanation, but that's why I placed it in a separate NOTE box.

File: components/learn/modules/ROOT/pages/on-upgrades.adoc (hunk 14 above)
Comment (spalladino) [reply components/learn/modules/ROOT/pages/on-upgrades.adoc:14]: Fixed identifier. I really like the phrase.

File: components/learn/modules/ROOT/pages/on-upgrades.adoc (hunk 38 above)
Comment (spalladino) [reply components/learn/modules/ROOT/pages/on-upgrades.adoc]: It will once https://github.com/OpenZeppelin/openzeppelin-sdk/pull/1366 is merged ;-)
